Fix #1251: Indicate that app is loading content for empty views

aforcier opened the pull request:
Fixes [#1251](https://github.com/wordpress-mobile/WordPress-Android/issues/1251). When opening a posts list that doesn't have any posts, a loading message is now shown when the app is loading data. This change also applies to pages, comments and media.

The loading message is replaced by either the actual content (posts, media, etc.) or the original "no content" message when the data has finished loading.

Previously the messages displayed were `No posts yet. Why not create one?`, `No comments` and `No media`.

Screen while fetching posts for the first time, before and after:

![posts-loading-before-after](https://cloud.githubusercontent.com/assets/9613966/5781188/9ac46660-9d7e-11e4-964d-c58a6ef6d186.png)

aforcier commented:
For the pages/posts loading message I left the quill-and-ink image in, but perhaps it's confusing and it would be better not to have any image or show a different loading resource?

maxme commented:
We should also have a "Network error" message in the following case:
User tries to load his posts/pages/comments/etc... for the first time and he gets a network error (airplane mode, server unavailable, bad connection,...) - "Loading..." should become "Network error" and not the default "No posts yet. Why not create one?".

maxme commented:
> For the pages/posts loading message I left the quill-and-ink image in, but perhaps it's confusing and it would be better not to have any image or show a different loading resource?

Yes, that could be great to have an animation on these loading

tonyr59h commented:
Looks good but I can't seem to find the branch on git to test it. Has it been pushed to origin?

maxme commented:
> Looks good but I can't seem to find the branch on git to test it. Has it been pushed to origin?

This branch comes from Alex' own repo:

```
git checkout -b ssue/1251-first-empty-views-loading develop
git pull https://github.com/aforcier/WordPress-Android.git issue/1251-first-empty-views-loading
```

tonyr59h commented:
D'oh! Didn't even notice:
> aforcier:issue/1251-first-empty-views-loading

aforcier commented:
Made lots of changes, here's a summary:

Added three error messages to the empty view that can be displayed after loading:
* Network Error in case of connectivity issues
* Permission Error ("You don't have permission to view or edit x") if the user is not authorized to view the content (pages, posts, comments, or media)
* Generic Error ("x couldn't be refreshed at this time" or "Something went wrong...") - the case and message for these already existed as a toast, and now it's also displayed in the empty view

Fixed the following bugs that became more obvious after my changes:
* Fixed an issue where the `SwipeRefreshLayout` would not display when swiping to refresh the posts/pages list and media grid when the view was empty (@tonyr59h pointed this out on Slack)
* Fixed a bug in `MediaGridFragment` where swiping to refresh or changing blogs with "Custom Date..." selected (or while searching) would show the `SwipeRefreshLayout` and never dismiss it
* Changed the comments list to display a "No comments" screen automatically when deleting the last comment (it used to show a blank screen)

Here are examples of the Loading, Network Error, Permission Error and Generic Error screens:

![1251-all-screens-resized](https://cloud.githubusercontent.com/assets/9613966/5823806/bc63d134-a0ad-11e4-888a-1025811fce23.png)

All of the different activities (posts/pages, comments and media) will now show the loading animation in the following cases: swiping, blog changing and on login (as shown in the first image). There is one case in which they don't show the animation (see known issues below).

Known issues:
* When opening any screen that refreshes automatically, the refresh animation (the `SwipeRefreshLayout`) doesn't show up. This happens whether there's content to be displayed on that screen or not. This seems to be a known issue with the android v4 library. I've opened a ticket [here](https://github.com/wordpress-mobile/WordPress-Android/issues/2220).

maxme commented:
It looks like we have a background color inconsistency on these views (grey / white / light blue).

tonyr59h commented:
> For the pages/posts loading message I left the quill-and-ink image in, but perhaps it's confusing and it would be better not to have any image or show a different loading resource?

I don't mind removing the image for loading. If we're going to have it display when there are no posts and remove it when the user swipes to refresh it would look better with an animation. Right now it disappears and appears instantaneously, looks a little jarring.

Maybe fade it out and move the text up as the user is pulling the swipe to refresh control down?

tonyr59h commented:
> Fixed an issue where the SwipeRefreshLayout would not display when swiping to refresh the posts/pages list and media grid when the view was empty

Working well on my test devices, good to see the consistency.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentsListFragment.java:
```diff
@@ -131,13 +154,21 @@ public void removeComment(Comment comment) {
+        mEmptyViewMessageType = MessageType.getEnumFromString(extras.getString(CommentsActivity.KEY_EMPTY_VIEW_MESSAGE));
+
         if (!NetworkUtils.isNetworkAvailable(getActivity())) {
+            updateEmptyView(MessageType.NETWORK_ERROR);
```
This works as expected when switching from another page to the Comments page, displaying Network Error. There's an edge case when the user switches blogs where it will display 'Comments couldn't be refreshed at this time' when it should be Network Error. I can post a video if this needs clarification.

Look into the UpdateCommentsTask, onPostExecute is toasting and displaying a GENERIC_ERROR because it doesn't check for a network connection.

aforcier replied:
Ah yes I see the issue, Media also does it.

tonyr59h replied:
:+1: Thanks for testing the other pages, slipped by my review.

tonyr59h commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -75,6 +75,7 @@
     <string name="upload">Upload</string>
     <string name="learn_more">Learn more</string>
     <string name="posting_post">Posting \"%s\"</string>
+    <string name="network_error">Network Error</string>
```
Might be worth it to consider using @string/no_network_message in lieu of adding a new resource. It seems to cover all the cases @string/network_error is used for.

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/posts/PostsListFragment.java:
```diff
@@ -49,9 +48,18 @@
     private boolean mCanLoadMorePosts = true;
     private boolean mIsPage, mShouldSelectFirstPost, mIsFetchingPosts;
 
+    private enum MessageType {
```
There are three definitions of MessageType, nearly identical. Any way we can consolidate?

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaGridFragment.java:
```diff
@@ -437,20 +476,53 @@ public void setFilterVisibility(int visibility) {
-            mEmptyView.setVisibility(View.VISIBLE);
-            if (messageId != -1) {
-                mEmptyViewTitle.setText(getResources().getString(messageId));
+    private void showEmptyViewWithMessage(MessageType messageType) {
```
The other implementations of updateEmptyView handle everything contained in showEmptyViewWithMessage. For consistency and simplicity I'd recommend doing the same here.

aforcier replied:
Oops, I thought I'd already refactored that part! I'll fix that.

aforcier commented:
Updates:

* Added an animation sequence to the Posts/Pages list: https://cloudup.com/cDuKvxJ46t5 (swiped to refresh at 00:04)
* Fixed issue where Media and Comments would show `GENERIC_ERROR` instead of `NETWORK_ERROR` when switching blogs while not connected
* Changed the `Network Error` message to the pre-existing `There is no network available` (I've updated the screenshot above)
* Some tidying up

In progress:
* Cleaning up some of the animation logic for better readability

tonyr59h commented:
Looking good, @aforcier! I tested it on my Galaxy S3 and, sans the alpha animations, it's working well.

I'm actually running into the alpha animation issue on some work I'm doing right now. Using
>ObjectAnimator.ofFloat(imageView, View.ALPHA, 0.25f, 1.0f).setDuration(duration).start()

doesn't seem to work on 4.1.X Android (at least not my S3 with TouchWiz). Seems to work fine if I use
>Animation alpha = new AlphaAnimation(0.25f, 1.0f);
>alpha.setDuration(duration);
>imageView.startAnimation(alpha);

but it's up to you if you want to support it, it still looks good without the alpha.

aforcier commented:
Did some refactoring to address @tonyr59h's comments:

* Used an `Enum` to track animation stages instead of incrementing an `int`
* Clarified the math for delaying consecutive animations
* Moved constant definitions for consistent style

tonyr59h commented:
:+1:

maxme commented:
Before merging this into develop, I have a few comments:

* Rename MessageType to a less generic name (EmptyViewMessageType ?).
* Decouple EmptyViewAnimationHandler from PostsListFragment and move it to its own file. (and move AnimationStage there).
* I noticed we use "Loading" and "Fetching" in different places. We should use the same word everywhere on these empty views.

aforcier commented:
Updates (based on @maxme's comments):
* Renamed the `MessageType` `enum` to `EmptyViewMessageType`
* Extracted `EmptyViewAnimationHandler` from `PostsListFragment` into a separate file along with `AnimationStage`
* Changed "Loading" empty view messages to "Fetching"

maxme commented:
:shipit:
